framework/email/tasks.py

```diff
@@ -4,50 +4,84 @@
-                username=None, password=None, mail_server=None):
+                username=None, password=None, mail_server=None, categories=None):
     """Send email to specified destination.
     Email is sent from the email specified in FROM_EMAIL settings in the
     settings module.
 
+    Uses the Sendgrid API if ``settings.SENDGRID_API_KEY`` is set.
+
     :param from_addr: A string, the sender email
     :param to_addr: A string, the recipient
     :param subject: subject of email
     :param message: body of message
+    :param tuple categories: Categories to add to the email using SendGrid's
+        SMTPAPI. Used for email analytics.
+        See https://sendgrid.com/docs/User_Guide/Statistics/categories.html
+        This parameter is only respected if using the Sendgrid API.
+        ``settings.SENDGRID_API_KEY`` must be set.
 
     :return: True if successful
     """
-    username = username or settings.MAIL_USERNAME
-    password = password or settings.MAIL_PASSWORD
-    mail_server = mail_server or settings.MAIL_SERVER
-
     if not settings.USE_EMAIL:
         return
-    if login and (username is None or password is None):
-        logger.error('Mail username and password not set; skipping send.')
-        return
+    if settings.SENDGRID_API_KEY:
+        return send_with_sendgrid(
+            from_addr=from_addr,
+            to_addr=to_addr,
+            subject=subject,
+            message=message,
+            mimetype=mimetype,
+            categories=categories
+        )
+    else:
+        username = username or settings.MAIL_USERNAME
+        password = password or settings.MAIL_PASSWORD
+        mail_server = mail_server or settings.MAIL_SERVER
 
-    msg = MIMEText(message, mimetype, _charset='utf-8')
-    msg['Subject'] = subject
-    msg['From'] = from_addr
-    msg['To'] = to_addr
+        if login and (username is None or password is None):
+            logger.error('Mail username and password not set; skipping send.')
+            return
 
-    s = smtplib.SMTP(mail_server)
-    s.ehlo()
-    if ttls:
-        s.starttls()
+        msg = MIMEText(message, mimetype, _charset='utf-8')
+        msg['Subject'] = subject
+        msg['From'] = from_addr
+        msg['To'] = to_addr
+
+        s = smtplib.SMTP(mail_server)
         s.ehlo()
-    if login:
-        s.login(username, password)
-    s.sendmail(
-        from_addr=from_addr,
-        to_addrs=[to_addr],
-        msg=msg.as_string()
-    )
-    s.quit()
-    return True
+        if ttls:
+            s.starttls()
+            s.ehlo()
+        if login:
+            s.login(username, password)
+        s.sendmail(
+            from_addr=from_addr,
+            to_addrs=[to_addr],
+            msg=msg.as_string()
+        )
+        s.quit()
+        return True
+
+def send_with_sendgrid(from_addr, to_addr, subject, message, mimetype='html', categories=None, client=None):
```

Comment:
Consider prefixing with `_` to soft-indicate private API (this function ain't a task).